src/test/lib/optimizer/strategy/null_scan_removal_rule_test.cpp

```diff
@@ -0,0 +1,103 @@
+  // The rule can't apply on a node that is not of type Predicate.
+  const auto actual_lqp = apply_rule(rule, mock_node);
+  const auto expected_lqp = mock_node->deep_copy();
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(NullScanRemovalRuleTest, PredicateIsNotNullExpression) {
+  // The rule can't apply on a predicate that is not a `null expression`.
+  const auto input_lqp = PredicateNode::make(equals_(mock_node_column, 42), mock_node);
+  const auto actual_lqp = apply_rule(rule, input_lqp);
+  const auto expected_lqp = input_lqp->deep_copy();
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(NullScanRemovalRuleTest, PredicateConditionIsNotNull) {
+  // The rule can't apply on a predicate which condition is not `is not null`.
+  const auto input_lqp = PredicateNode::make(is_null_(mock_node_column), mock_node);
+  const auto actual_lqp = apply_rule(rule, input_lqp);
+  const auto expected_lqp = input_lqp->deep_copy();
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(NullScanRemovalRuleTest, PredicateOperandIsNotLQPColumnExpression) {
+  // The rule can't apply where the predicate operand is not a `LQP Column expression`.
+  const auto input_lqp = PredicateNode::make(is_not_null_(42), mock_node);
+  const auto actual_lqp = apply_rule(rule, input_lqp);
+  const auto expected_lqp = input_lqp->deep_copy();
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(NullScanRemovalRuleTest, LQPColumnOriginalNodeIsNotStoredTableNode) {
+  // The rule can't apply where the original node of the `LQP Column expression` is not a `storage table node`.
+  const auto input_lqp = PredicateNode::make(is_not_null_(mock_node_column), mock_node);
+  const auto actual_lqp = apply_rule(rule, input_lqp);
+  const auto expected_lqp = input_lqp->deep_copy();
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(NullScanRemovalRuleTest, TableColumnDefinitionIsNullable) {
+  // The rule can't apply, if the column is not nullable
+  const auto input_lqp = PredicateNode::make(is_not_null_(nullable_table_node_column), nullable_table_node);
+  const auto actual_lqp = apply_rule(rule, input_lqp);
+  const auto expected_lqp = input_lqp->deep_copy();
+
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(NullScanRemovalRuleTest, TableColumnDefinitionIsNotNullable) {
+  // All needed conditions to remove the node are matched:
+  // 1. The node must be of type Predicate
+  // 2. The predicate must be a `null expression`
+  // 3. The predicate condition must be `is not null`
+  // 4. The predicate operand needs to be an `LQP Column expression`
+  // 5. The original node of the `LQP Column expression` needs to be a not defective `storage table node`
+  // 6. The column (referenced by the `LQP Column expression`) is not nullable
+  const auto input_lqp = PredicateNode::make(is_not_null_(table_node_column), table_node);
+  const auto actual_lqp = apply_rule(rule, input_lqp);
+  const auto expected_lqp = table_node->deep_copy();
```

Comment:
Why are you performing a deep copy here?

Reply:
If I am not doing a deep copy I get the error that comparing an LQP with itself is always true.

Reply:
Ah, I remember. Yes, in this case the deep_copy is necessary, but it should still happen BEFORE you apply the rule.

Reply:
I already changed the code so that the deep copy is done before applying the rule. I am not sure why GitHub is not showing that the file view is outdated.